README.md
# Generate Ren'Py Scripting

![README Image](./images/plugin.png)

Generate Ren'Py Scripting is a Python plugin for Krita. GRS outputs a block of Ren'Py scripting to display the images of the active Krita document as they appear on the canvas, working in tandem with the [krita-batch-exporter](https://github.com/GDQuest/krita-batch-exporter). The output file is automatically opened so that you may immediately copy and paste the text into your Ren'Py project. The idea is to make working between Ren'Py and Krita as efficient as possible!

# How to Use

 1. Determine the image export scale you are going to use. The included Scale Percentage Size Calculator will tell you the dimensions of the image at the given scale; for Ren'Py, you'll likely want to get it close to 1920x1080 pixels.
 2. Batch export the image's elements using KBE (krita-batch-exporter). This means that the layers you wish to export must follow the syntax defined in the [KBE manual](https://github.com/GDquest/krita-batch-exporter/blob/master/batch_exporter/Manual.md); this program's features will not work with incorrectly formatted layer names! The layer name (the actual name substring) will be used as the image name in the Ren'Py script.
 3. (Optional) Customize the name of the output file (rpblock.txt by default). Do so if you wish to make and save multiple output files, though the practical use is to have rpblock.txt as a temporary file akin to scratch paper.
 4. Choose your output type.
	1. If you need `pos (x, y)` scripting, press `pos (x, y)`.
    2. If you need `align (x, y)` or `xalign x yalign y` scripting:
       1. Choose the number of evenly-distributed spacings to use. <br>The Rule of Thirds toggle has the same effect as choosing 4 spacings [0.0, 0.333, 0.666, 1.0].
       2. Press `align (x, y)` or `xalign x yalign y`.
 5. (Optional) Use the Additional ATL system to add an `at function()` statement.
	 1. ATL functions may be written with a layer with the name `ATL <name of layer to target>`.
	 2. The function must be given in the format `f=<function(parameter 1, parameter 2, etc.)>`. `func` and `function` also work as tags.
	 3. If `currX` and `currY` are used in the string, the output will use the bounding box top left corner coordinates of the layer's contents, scaled to the lowest given resolution. For example, if the target layer has the metadata `s=100,50`, `currX` and `currY` will be calculated for 50% the image's size, and then written into the exported text. Any other variable that you type in will simply be carried over as-is.
	 4. Property tags such as `alpha` and `subpixel` are supported; see the manual for details, including the full list.
 6. After you choose a save location, the exported file will be opened in the program default for its type automatically. Copy and paste its text where you need it in your Ren'Py script. I like to save to Desktop so that I could find and delete the file afterwards easily.
 7. Make changes to your Ren'Py script as necessary. GRS gets you started with the basic image display template and can be used to add an `at function()` statement, but any further ATL statements still need to be declared manually!

For more information with examples, see the [manual](https://github.com/SeanHRN/generate-renpy-scripting/blob/master/manual.md).

# Download and Installation
Quickest Way via Python Plugin Manager (from Web):
Open Krita and go to `Tools -> Scripts -> Import Python Plugin from Web`.
Enter as the download URL: https://github.com/SeanHRN/generate-renpy-scripting

Alternatively:
Download the [zip or tar.gz](https://github.com/SeanHRN/generate-renpy-scripting/releases).

Easy Install via Python Plugin Manager (from File):
Open Krita and go to `Tools -> Scripts -> Import Python Plugin from File`. Select `generateRenpyScripting_x.x.zip`.

Long Way:
Unzip `generate-renpy-scripting-x.x.x.zip`. Place `generateRenpyScripting` (the folder) and `generateRenpyScripting.desktop` in the pykrita directory.
You may find the location of the pykrita folder by opening Krita and clicking
 `Settings -> Manage Resources -> Open Resource Folder`.

Proceed to Enabling GRS in Krita.

# Enabling GRS in Krita
1. Open Krita.
2. Go to `Settings -> Configure Krita -> Python Plugin Manager`.
3. `Generate Ren'Py Scripting` will be in the list; check its box.
4.  Restart Krita.

# Features To Possibly Implement
  - Compatibility with the Batch Exporter's custom output path feature.
  - GUI text bar to notify the user of success/failure in export to match Krita Batch Exporter's design.<br> Currently, a separate window opens if the user tries to run the plugin when a Krita file hasn't been opened, and the only notification of success is the automatic file opening.

# Known Issues
 - The calculator's dimensions don't automatically update when the canvas's size changes,
 for example, upon cropping. I don't know of a signal to handle that.
 - The main example in the manual is outdated because it was made before the Batch Exporter allowed non-whole numbers \
for scale percentage. I will make a new example.

# Feedback
Do you have suggestions on how to make the plugin better? Are there any commonly-used templates that you want to see as export options? Let me know here.

# License
Generate Ren'Py Scripting is released under the GNU General Public License (version 3 or later).


changelog.md
# Change Log

# v1.8
 - Fixed the issue described in v1.6.1 for real.
 - Expanded the alpha system to a full property list system.
 - Updated the html manual (renamed from grs_manual.html to Manual.html).
 - Updated the md manual.

# v1.7
 - New export option: `at setPos(x, y)`. Since there are 4 export buttons,
the export buttons are now on a grid layout.
 - Removed newline between `pause` and the rest of the block.

# v1.6.1
 - Fixed issue in which the image renamer would copy over an empty directory \
during this case (using smallest scale = 25% as an example):
     1. Run the image renamer. The export subdirectory is `x0.25`.
     2. Remove all the contents of `x0.25` (for example, if you were to \
cut and paste those images elsewhere).
     3. Run the image renamer again. `x0.25` is repopulated with the images, \
but also the empty folder `x0.25\x0.25`.

# v1.6
 - Added Additional ATL System.

# v1.5.3
 - Fixed error from using `t=false`.

# v1.5.2
 - Added handling for when user tries to use the image renamer when none of the layer names is valid
 - Added handling for when user tries to use the image renamer when a document isn't active

# v1.5.1
 - Changed an instance of int() to float() for scale to correct an exporting issue.
 - Altered the smallest scale calculation to accommodate 200% (from 100%).

# v1.5
 - Calculator Rework
     - Now uses floats because the Batch Exporter will soon allow non-whole number scale percentages.
     - Modifier key increment system: Hold a key to change the scale box's arrow increments.
         - Alt: 0.01%
         - Shift: 0.1%
         - Default: 1.0%
         - Ctrl: 10%
     - Width and Height numbers are now in editable text boxes.
         - Edit Width: Scale and Height will automatically update.
         - Edit Height: Scale and Width will automatically update.
     - Scale limit increased from 100% to 200% to allow dimension calculation for small upscaling.
     - Added a tooltip.
 - Added [xalign x yalign y] as an export option.
 - Layer name parsing now uses a dictionary system to be more \
robust and better for additional tag implementations.
# v1.4
 - `Rename Batch-Exported Files` is now a button instead of a checkbox.
 - `Rename Batch-Exported Files` now works with layers that were exported as groups into folders.
 - Minor code improvements: `"/"` -> `os.sep` and `min()` instead of a loop.

# v1.3
 - Fixed size parsing issue.
 - New feature: `Rename Batch-Exported Files` (until I have a more accurate name): \
 Copies the smallest scale batch-exported images to a folder where they don't have \
 the Batch Exporter's `_x[scale]` suffix in their names.

# v1.1.2
 - Added the in-Krita manual html file.

# v1.1
 - Scale Percentage Size Calculator properly updates the dimensions when a \
new document is loaded, when the user switches between active documents, \
and when all documents are closed.
 - Several GUI tweaks:
     - The export options are on a single line.
     - Instances of `pos(x, y)` and `align(x, y)` have been given a space between pos/align and the left parenthesis to look a bit nicer and be consistent with the output style: `pos (x, y)` and `align (x, y)`
     - `Scale Dimensions` is now capitalized to be consistent with the rest of the text.


generateRenpyScripting/generateRenpyScripting/generateRenpyScripting.py
from krita import *
import sys
import math
from pathlib import Path
import os
from os.path import join, exists, dirname
import webbrowser
from PyQt5.QtWidgets import *
from sys import platform
import subprocess
import shutil
import re
import json
from collections import defaultdict

KI = Krita.instance()
open_notifier = KI.notifier()
open_notifier.setActive(True)
close_notifier = KI.notifier()
close_notifier.setActive(True)

default_outfile_name = "rpblock.txt"
indent = 4
decimal_place_count = 3
transform_properties = {
                        "rotate", "rotate_pad", "transform_anchor",
                        "zoom", "xzoom", "yzoom", "nearest", "alpha",
                        "additive", "around", "alignaround", "crop",
                        "subpixel", "delay", "events", "xpan", "ypan",
                        "xtile", "ytile", "matrixcolor", "blur"
                        }

def closestNum(num_list, value):
    return num_list[min(range(len(num_list)), key = lambda i: abs(num_list[i]-value))]

def truncate(number, digit_count):
    step = 10.0 ** digit_count
    return math.trunc(step * number) / step

def generateSpaces(spacing_count):
    step = 1.0 / (spacing_count - 1)
    spacing_list = []
    for i in range(spacing_count):
        spacing_list.append(truncate(i*step, decimal_place_count))
    return spacing_list

def convertKeyValue(input_dict_value):
    """
    convertKeyValue removes characters from a dictionary string that
    should not be printed. The input is a key, and the output is a string.
    """
    output = str(input_dict_value)
    return output.translate(str.maketrans('','','[]\''))

def parseLayers(layer, layer_list, coordinates_list, centers_list):
    """
    parseLayers() gets the data from the layers.
    """
    if layer.visible() == True:
        layer_sublist = []
        coordinates_sublist = []
        centers_sublist = []
        lower_n = layer.name().lower()
        if lower_n.find(" e=") != -1:
            coord_x, coord_y, center_point = 0, 0, [0,0]
            if lower_n.find(" t=false") == -1 and lower_n.find(" t=no") == -1:
                coord_x = layer.bounds().topLeft().x()
                coord_y = layer.bounds().topLeft().y()
                center_point = layer.bounds().center()
                centers_list.append([center_point.x(), center_point.y()])
            else:
                currentDoc = KI.activeDocument()
                width = currentDoc.width()
                height = currentDoc.height()
                centers_list.append([round(width/2), round(height/2)])
            layer_list.append(layer)
            coordinates_list.append([coord_x, coord_y])
        elif layer.type() == "grouplayer":
                for child in layer.childNodes():
                    parseLayers(child, layer_sublist, coordinates_sublist, \
centers_sublist)
        layer_list.extend(layer_sublist)
        coordinates_list.extend(coordinates_sublist)
        centers_list.extend(centers_sublist)

def calculateAlign(data_list, centers_list, spacing_count):
    """
    calculateAlign converts the pos(x,y) coordinates
    in the data list into align(x,y) coordinates
    by comparing the center point of each image
    to a finite set of values from 0.0 to 1.0.
    """
    width, height = 1, 1
    currentDoc = KI.activeDocument()
    if currentDoc != None:
        width = currentDoc.width()
        height = currentDoc.height()

    spacing_list = generateSpaces(spacing_count)
    new_data_list = []
    for d, c in zip(data_list, centers_list):
        xalign = closestNum(spacing_list, (c[0] / width))
        yalign = closestNum(spacing_list, (c[1] / height))
        new_data_list.append(tuple((d[0],xalign,yalign,d[3])))
    return new_data_list

def getData(button_num, spacing_count):
    """
    Uses a dictionary system to parse the layer data
    and apply changes to coordinates.
    """
    file_open = False
    data_list = []
    layer_list = []
    contents = []
    all_coords = []
    all_centers = []
    layer_dict = defaultdict(dict)
    test_list = []
    currentDoc = KI.activeDocument()
    if currentDoc != None:
        file_open = True
        root_node = currentDoc.rootNode()
        for i in root_node.childNodes():
            parseLayers(i, layer_list, all_coords, all_centers)
    for l in layer_list:
        contents = l.name().split(" ")
        layer_dict[l.name()]["actual name"] = contents[0]
        for c in contents[1:]:
            if "=" in c:
                category_data = c.split("=") # 0: Category string, 1: Data string
                category_data_list = category_data[1].split(",")
                layer_dict[l.name()][category_data[0]] = category_data_list
    for layer, coord_indv in zip(layer_list, all_coords):
        if "m" in layer_dict[layer.name()]:
            for i in layer_dict[layer.name()]["m"]:
                margin_list = [int(i) for i in layer_dict[layer.name()]["m"]]
                coord_indv[0] -= max(margin_list)
                coord_indv[1] -= max(margin_list)
        if "s" in layer_dict[layer.name()]:
            size_list = [float(i) for i in layer_dict[layer.name()]["s"]]
            x = round(coord_indv[0] * (min(size_list)/100))
            y = round(coord_indv[1] * (min(size_list)/100))
            data_list.append(tuple((layer_dict[layer.name()]["actual name"], \
x, y, size_list)))
    if button_num == 2 or button_num == 3:
        data_list = calculateAlign(data_list, all_centers, spacing_count)
    return file_open, data_list

def getATL(curr_node):
    """
    getATL() checks for ATL information in invisible layers.
    Since spaces are used to split the contents, I don't think
    the function statement could allow any spaces.

    Tags that are not recognized from the list transform_properties
    are recorded and put in the dictionary invalid_dict.
    """
    ATL_dict = defaultdict(dict)
    invalid_dict = defaultdict(dict)
    for i in curr_node.childNodes():
        if i.visible() == False and i.name().upper().startswith("ATL"):
            if i.type() == "grouplayer":
                ATL_dict.update(getATL(i))
            elif i.type() == "paintlayer":
                contents = i.name().split(" ")
                invalid_list = []
                for c in contents[2:]:
                    if "=" in c:
                        ATL_data = c.split("=") #0: Tag string, 1: Data String
                        ATL_data_list = ATL_data[1]
                        ATL_dict[contents[1]][ATL_data[0]] = ATL_data_list
                        if ATL_data[0] not in transform_properties:
                            invalid_list.append(ATL_data[0])
                if len(invalid_list) != 0:
                    invalid_dict[contents[1]] = invalid_list

    return ATL_dict, invalid_dict

def getATLFunction(input_string, input_layer_tuple, input_data):
    """
    The regex is a thing from stack overflow to find
    'comma that is followed by a char that is not a space'.
    """
    modified_string = input_string
    modified_string = re.sub('(,(?=\S)|:)', ", ", modified_string)
    for i in input_data:
        if i[0] == input_layer_tuple[0]:
            if "currX" in modified_string:
                modified_string = modified_string.replace("currX", str(input_layer_tuple[1]))
            if "currY" in modified_string:
                modified_string = modified_string.replace("currY", str(input_layer_tuple[2]))
    return modified_string

def writeData(input_data, path, format_num):
    ATL_dict = {}
    currentDoc = KI.activeDocument()
    if currentDoc != None:
        ATL_dict, invalid_dict = getATL(currentDoc.rootNode())
    outfile = open(path, "w")
    outfile.write("\n")
    prefix = "pos"
    if format_num == 2:
        prefix = "align"

    # for debugging
#    outfile.write(str(ATL_dict) + "\n")
#    outfile.write(str(invalid_dict) + "\n")

    for d in input_data:
        at_statement = ""
        ATL = ""
        property_dict = {}
        for t in transform_properties:
            property_dict[t] = None
        optional_colon = ":"
        no_property_block = True
        if d[0] in ATL_dict:
            for key in property_dict:
                if key in ATL_dict[d[0]]:
                    no_property_block = False
                    property_dict[key] = ATL_dict[d[0]][key]
            for f in ["f", "func", "function"]:
                if f in ATL_dict[d[0]]:
                    ATL = getATLFunction(ATL_dict[d[0]][f], d, input_data)
                    break
        if format_num == 4:
            at_statement = " at setPos(" + str(d[1]) + ", " + str(d[2]) + ")"
            if no_property_block:
                optional_colon = ""
        elif ATL:
            at_statement = " at " + ATL

        outfile.write(f"{' ' * indent}show {d[0]}{at_statement}{optional_colon}\n")
        if format_num == 1 or format_num == 2:
            outfile.write(f"{' ' * (indent * 2)}{prefix} ({str(d[1])}, {str(d[2])})\n")
        elif format_num == 3:
            outfile.write(f"{' ' * (indent * 2)}xalign {str(d[1])} yalign {str(d[2])}\n")
        for key in property_dict:
            if property_dict[key] is not None:
                outfile.write(f"{' ' * (indent * 2)}{key} {property_dict[key]}\n")
        if ATL and format_num == 4:
            outfile.write(f"{' ' * (indent * 2)}{ATL}\n")

    outfile.write(f"{' ' * indent}pause")

    if invalid_dict:
        chars_to_remove = ['[', ']', '\'']
        outfile.write(f"\n# Suppressed unrecognized transform property/properties:")
        for layer in invalid_dict:
            outfile.write(f"\n#{' ' * indent}{invalid_dict[layer]} in layer \"{layer}\"")
    outfile.close()

def renameRecursion(dir_name, export_dir_name, suffix, folder_name):
    """
    Helper function for recursiveRenameStart()
    dir_name:        the directory to be copied.
    export_dir_name: the directory into which the files are copied.
    suffix:          the part of the file name string that is to be removed.
    folder_name:     name of the top folder of the copy directory;
                     it's needed to prevent infinite looping.
    """
    for filename in os.listdir(dir_name):
        f = os.path.join(dir_name, filename)
        if filename.find(suffix) != -1:
            if os.path.isfile(f):
                exp_fname, exp_ext = os.path.splitext(filename)
                exp_fname = exp_fname[:exp_fname.find(suffix)]
                exp_fname += exp_ext
                dst = os.path.join(export_dir_name, exp_fname)
                shutil.copy(f, dst)
        elif os.path.isdir(f):
            if filename == folder_name:
                continue
            else:
                sub_export_dir_name = os.path.join(export_dir_name, filename)
                Path(sub_export_dir_name).mkdir(parents=True, exist_ok=False)
                renameRecursion(f, sub_export_dir_name, suffix, folder_name)

def recursiveRenameStart(data_list):
    """
    Uses data_list to get the layer names and the scales.
    A folder named after the smallest scale is created.
    The images of that scale are copied over with the scale tag removed from
    their names.
    The data_list gives the scale list in numbers out of 100, so this function
    divides those numbers by 100 to get the multiplier for the folder name.
    Pre-requisite: KI.activeDocument() != None
    """
    dir_name = os.path.dirname(KI.activeDocument().fileName())
    dir_name = os.path.join(dir_name, "export")
    smallest_scale = 200.0
    for d in data_list:
        if min(d[3]) < smallest_scale:
            smallest_scale = min(d[3])
    suffix = "_@" + str(smallest_scale/100) + "x"
    new_folder_name = "x" + str(smallest_scale/100.0)
    export_dir_name = dir_name + os.sep + new_folder_name
    Path(export_dir_name).mkdir(parents=True, exist_ok=True)
    #    of = open(os.path.join(export_dir_name, "diagnostic.txt"), "w")
    #    of.write(dir_name + "\n" + export_dir_name + "\n" + suffix + "\n")
    #    of.close()
    renameRecursion(dir_name, export_dir_name, suffix, new_folder_name)

class CustomDoubleSpinBox(QDoubleSpinBox):
    """
    Customized double spin box with these modifier increments for arrow click:
    Alt: 0.01
    Shift: 0.1
    No Modifier: 1
    Ctrl: 10 (It appears to be default in Qt.)
    """
    def __init__(self, parent=None):
        super(CustomDoubleSpinBox, self).__init__(parent)

    def stepBy(self, steps):
        modifiers = QApplication.keyboardModifiers()
        if modifiers == QtCore.Qt.AltModifier:
            QDoubleSpinBox.setSingleStep(self, 0.01)
        elif modifiers == QtCore.Qt.ShiftModifier:
            QDoubleSpinBox.setSingleStep(self, 0.1)
        else:
            QDoubleSpinBox.setSingleStep(self, 1.0)
        QDoubleSpinBox.stepBy(self, steps)

class GenerateRenpyScripting(DockWidget):
    title = "Generate Ren'Py Scripting"

    def __init__(self):
        super().__init__()
        self.setWindowTitle(self.title)
        self.createInterface()
        self.mainWindow = None
        open_notifier.imageCreated.connect(self.updateScaleCalculation)
        open_notifier.imageCreated.connect(self.initiateScaleCalculation)
        close_notifier.viewClosed.connect(self.wasLast)

    def initiateScaleCalculation(self):
        self.mainWindow = KI.activeWindow()
        self.mainWindow.activeViewChanged.connect(self.updateScaleCalculation)

    def wasLast(self):
        """
        At the moment a view is closed, the view is still part of the window's
        view count, so when the final view is closed, Window.views() is 1.
        Set the dimensions back to 0x0.
        """
        if len(KI.activeWindow().views()) == 1:
            self.wipeScaleCalculation()

    def showErrorMessage(self, toPrint):
        msg = QMessageBox()
        msg.setText(toPrint)
        msg.exec_()

    def createInterface(self):
        export_label = QLabel("Export")
        export_label.setToolTip("Export a file with a block of Ren'Py scripting\
for a quick copy and paste.")

        pos_button = QPushButton("pos (x, y)")
        pos_button.clicked.connect(lambda: self.process(1))

        align_button = QPushButton("align (x, y)")
        align_button.clicked.connect(lambda: self.process(2))

        xalignyalign_button = QPushButton("xalign x yalign y")
        xalignyalign_button.clicked.connect(lambda: self.process(3))

        atSetPos_button = QPushButton("at setPos(x, y)")
        atSetPos_button.clicked.connect(lambda: self.process(4))

        self.spacing_slider = QSlider(Qt.Horizontal, self)
        self.spacing_slider.setGeometry(30, 40, 200, 30)
        self.spacing_slider.setRange(2, 9)
        self.spacing_slider.setValue(9)
        self.spacing_slider.setFocusPolicy(Qt.NoFocus)
        self.spacing_slider.setPageStep(1)
        self.spacing_slider.setTickInterval(1)
        self.spacing_slider.setTickPosition(QSlider.TicksBelow)
        self.spacing_slider.valueChanged[int].connect(self.updateSpacingValue)
        self.spacing_label = QLabel("align (x, y) Spacing Count: ")
        self.spacing_label.setToolTip("Choose number of evenly-distributed \
spaces to use for align(x, y).")
        self.spacing_number_label = QLabel(f"{self.spacing_slider.value()}")
        self.spacing_number_label.setAlignment(Qt.AlignVCenter)
        self.rule_of_thirds_check = QCheckBox("Rule of Thirds")
        self.rule_of_thirds_check.setToolTip("Set align(x, y) \
statements to Rule of Thirds intersections. This is equivalent to using 4 spaces.")
        self.rule_of_thirds_check.setChecked(False)

        scale_label = QLabel("Scale Percentage Size Calculator")
        scale_label.setToolTip("Check the image's dimensions \
at the given scale.\nHold Alt to increment by 0.01%.\nHold Shift to increment \
by 0.1%.\nHold Ctrl to edit by 10%.")
        self.scale_box_percent = CustomDoubleSpinBox(self)
        self.scale_box_percent.setRange(0.0, 200.0)
        self.scale_box_percent.setValue(100.0)
        self.scale_box_percent.valueChanged[float].connect(self.calculatorScaleChanged)
        scale_text = QLabel("% Scale Dimensions:")
        self.onlyDouble = QDoubleValidator()
        width_label = QLabel("Width:")
        self.calculator_width = QLineEdit(self)
        self.calculator_width.setValidator(self.onlyDouble)
        self.calculator_width.setText("0.0")
        self.calculator_width.textEdited[str].connect(self.calculatorWidthEdited)
        height_label = QLabel("Height:")
        self.calculator_height = QLineEdit(self)
        self.calculator_height.setValidator(self.onlyDouble)
        self.calculator_height.setText("0.0")
        self.calculator_height.textEdited[str].connect(self.calculatorHeightEdited)

        rename_button = QPushButton("Rename Batch-Exported Files")
        rename_button.clicked.connect(lambda: self.renameClicked())
        rename_button.setToolTip("Option to save copies of the Batch-Exported \
files of the smallest scale without the size suffix, placed in a folder.")

        filename_label = QLabel("Output File")
        self.filename_line = QLineEdit()
        self.filename_line.setToolTip("Output file name with .[extension]")
        self.filename_line.setText(default_outfile_name)

        main_layout = QVBoxLayout()
        export_layout = QGridLayout()
        spacing_layout = QHBoxLayout()
        scale_layout = QHBoxLayout()

        main_layout.addWidget(export_label)
        export_layout.addWidget(pos_button, 0, 0)
        export_layout.addWidget(align_button, 1, 0)
        export_layout.addWidget(xalignyalign_button, 1, 1)
        export_layout.addWidget(atSetPos_button, 0, 1)
        export_layout.setContentsMargins(0, 4, 0, 2) #left top right bottom
        main_layout.addLayout(export_layout)

        main_layout.addWidget(self.spacing_label)
        spacing_layout.setContentsMargins(0,0,0,0)
        spacing_layout.addWidget(self.spacing_number_label)
        spacing_layout.addWidget(self.spacing_slider)
        spacing_layout.addWidget(self.rule_of_thirds_check)
        main_layout.addLayout(spacing_layout)

        main_layout.addWidget(scale_label)
        self.scale_box_percent.setGeometry(0, 0, 10, 10)
        scale_layout.addWidget(self.scale_box_percent)
        scale_layout.addWidget(scale_text)
        scale_layout.addWidget(width_label)
        scale_layout.addWidget(self.calculator_width)
        scale_layout.addWidget(height_label)
        scale_layout.addWidget(self.calculator_height)
        main_layout.addLayout(scale_layout)

        main_layout.addWidget(rename_button)
        main_layout.addWidget(filename_label)
        main_layout.addWidget(self.filename_line)
        main_layout.addStretch()

        mainWidget = QWidget(self)
        mainWidget.setLayout(main_layout)
        self.setWidget(mainWidget)

    def updateScaleCalculation(self):
        multiplier = float(self.scale_box_percent.value() / 100.0)
        currentDoc = KI.activeDocument()
        width, height = 0.0, 0.0
        if currentDoc != None:
            width = round((currentDoc.width() * multiplier), decimal_place_count)
            height = round((currentDoc.height() * multiplier), decimal_place_count)
        self.calculator_width.setText(f"{width}")
        self.calculator_height.setText(f"{height}")

    def wipeScaleCalculation(self):
        self.calculator_width.setText("0")
        self.calculator_height.setText("0")

    def calculatorScaleChanged(self):
        currentDoc = KI.activeDocument()
        multiplier = float(self.scale_box_percent.value() / 100.0)
        if currentDoc != None:
            width = round((float(currentDoc.width()) * multiplier), decimal_place_count)
            height = round((float(currentDoc.height()) * multiplier), decimal_place_count)
            if self.calculator_width.hasFocus() == False:
                self.calculator_width.setText(str(width))
            if self.calculator_height.hasFocus() == False:
                self.calculator_height.setText(str(height))

    def calculatorWidthEdited(self):
        """
        Difference between QLineEdit::textChanged() and QLineEdit::textEdited:
        textEdited() is emitted whenever the text is directly edited.
        textChanged() is emitted whenever the text is edited by any means.
        Use textEdited() to not have calculation loops.
        """
        currentDoc = KI.activeDocument()
        multiplier = 1.0
        if currentDoc != None:
            try:
                multiplier = float(self.calculator_width.text()) / currentDoc.width()
            except:
                multiplier = 1.0
            if self.scale_box_percent.hasFocus() == False:
                self.scale_box_percent.setValue(100 * multiplier)
            self.calculator_height.setText(f"{round(float(currentDoc.height() * multiplier), 2)}")

    def calculatorHeightEdited(self):
        """
        Difference between QLineEdit::textChanged() and QLineEdit::textEdited:
        textEdited() is emitted whenever the text is directly edited.
        textChanged() is emitted whenever the text is edited by any means.
        Use textEdited() to not have calculation loops.
        """
        currentDoc = KI.activeDocument()
        multiplier = 1.0
        if currentDoc != None:
            try:
                multiplier = float(self.calculator_height.text()) / currentDoc.height()
            except:
                multiplier = 1.0
            if self.scale_box_percent.hasFocus() == False:
                self.scale_box_percent.setValue(100 * multiplier)
            self.calculator_width.setText(f"{round(float(currentDoc.width() * multiplier), 2)}")

    def updateSpacingValue(self):
        self.spacing_number_label.setText(str(self.spacing_slider.value()))

    def process(self, button_num):
        spacing_count = 9 # Just a placeholder value that works.
        if self.rule_of_thirds_check.isChecked():
            spacing_count = 4
        else:
            spacing_count = self.spacing_slider.value()
        file_open_test_result, data = getData(button_num, spacing_count)
        if file_open_test_result == True:
            path = str(QFileDialog.getExistingDirectory(None, "Select a save location."))
            outfile_name = default_outfile_name
            if self.filename_line.text() != "":
                outfile_name = self.filename_line.text()
            path += os.sep + outfile_name
            writeData(data, path, button_num)
            outfile_exists = exists(path)
            if outfile_exists:
                webbrowser.open(path)
            else:
                push_message = "Failure: Open a Krita document."
                QMessageBox.information(QWidget(), "Generate Ren'Py Scripting", push_message)

    def renameClicked(self):
        """
        The getData() call here ignores the first value
        and gets the second value. The inputs [1, 9] don't really matter here.
        """
        data = getData(1, 9)[1]
        if KI.activeDocument() != None:
            if data:
                recursiveRenameStart(data)
            else:
                self.showErrorMessage("Error: No valid layer names detected. \
The renamer cannot run. Check your layer names!")
        else:
            self.showErrorMessage("Error: No active document!")

    def canvasChanged(self, canvas):
        pass

def registerDocker():
    Krita.instance().addDockWidgetFactory(DockWidgetFactory\
("generateRenpyScripting", DockWidgetFactoryBase.DockRight\
 , GenerateRenpyScripting))
